Added cacheAsBitmap with example and a fix or two

SmilyOrg opened the pull request:
Added cacheAsBitmap functionality to DisplayObject with an example, currently supported by Sprite/DisplayObjectContainer (DOC)
Refactored/deduplicated Sprite/DOC rendering code to use a common function
Fixed leftover pixel shifting code that broke pixel perfect texture mapping on Windows due to OpenGL
Fixed render to texture to work while rendering to stage
Disabled FBO check if not in debug

This depends on #414, so please wait on that.

bengarney commented on loom/engine/loom2d/l2dDisplayObject.h:
```diff
@@ -113,6 +113,20 @@ class DisplayObject : public EventDispatcher
+    // if false, the contents are re-cached at render time
+    bool cacheAsBitmapValid;
+    // true if caching is in progress, used for avoiding rendering to texture while rendering to texture 
+    static bool cacheAsBitmapInProgress;
```
Should be private or something?

SmilyOrg replied:
Oh probably.

bengarney commented on loom/graphics/gfxGraphics.h:
```diff
@@ -163,6 +165,23 @@ namespace GFX
     } GL_Context;
 
+// Represents graphics view frame properties that can change from frame buffer to frame buffer
+typedef struct GraphicsFrame {
```
Given we render "frames" maybe it would be better to call this GraphicsRenderTarget?

SmilyOrg replied:
Yeah I was conflicted on that front, GraphicsRenderTarget is probably better!

bengarney commented:
This looks pretty good. And I think it implicitly passes events through to children yes? Some nits, once fixed and we confirm it works on iOS/Android I think we are good!

bengarney commented:
Do we need to worry about allocating RTs that are pow2 in size?

SmilyOrg commented:
Yeah it implicitly passes events to children (only the rendering part is swapped). Will have to test on iOS/Android and check for pow2 issues.

SmilyOrg commented:
Okay, addressed nits. As for platforms, I tested it and the functionality works fine on all 4 platforms, however the example shows up as a black screen on iOS due to stageWidth/Height returning 0 in run(). I've tested the functionality by hardcoding the sizes and it appears to work, even without rounding to POT dimensions. Fixing stageWidth/Height is out of scope for this task, so I've added a card for it to be fixed at a later point.

bengarney commented on loom/graphics/gfxTexture.cpp:
```diff
@@ -1174,9 +1174,15 @@ void Texture::validate(TextureID id)
 void Texture::setRenderTarget(TextureID id)
 {
-    LOOM_PROFILE_SCOPE(textureSetRenderTarget);
+	// TODO Gamma correct rendering? Render is lighter than it's supposed to be
```
Is there a card for this?

SmilyOrg replied:
There is now :)

bengarney commented:
OK, nitted and I think ready to land!
